### PR TITLE
Improvements to UnconfirmedPool, OrphanPool, PendingPool, and ReorgPool

### PR DESCRIPTION
**Pull request by neonknight64**
## Description
- Added min_input_maturity, max_input_maturity and max_timelock_height member functions to transactions
- Removed "mut" requirements for some of the member functions of blockchain_database
- Separated OrphanPool and PendingPool into an Interface and Storage sections with RwLocks
- Changed OrphanPool to have a reference to the BlockchainDatabase so it can correctly query the UTXO set
- Added snapshot and calculate_weight functions to the UnconfirmedPool, OrphanPool, PendingPool, and ReorgPool
- Modified FeePriority to include the oldest maturity of the input UTXO
- Modified TimelockPriority to correctly use the Transactions lock_height and UTXO maturity
- Modified create_test_tx and create_test_block to create more validly constructed transactions and blocks
- Added create_test_input for creating valid transaction inputs and corresponding UTXOs

## Motivation and Context
These changes were needed for the implementation of the Mempool

## How Has This Been Tested?
The existing tests were updated.

## Types of changes

* [x] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [ x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/core/src/mempool/reorg_pool/reorg_pool.rs [hunk 1]
```diff
@@ -53,57 +56,76 @@ impl Default for ReorgPoolConfig {
...
+            pool_storage: RwLock::new(ReorgPoolStorage::new(config)),
         }
     }
 
     /// Insert a new transaction into the ReorgPool. Published transactions will have a limited Time-to-live in the
     /// ReorgPool and will be discarded once the Time-to-live threshold has been reached.
-    pub fn insert(&mut self, tx: Transaction) {
-        let tx_key = tx.body.kernels[0].excess_sig.clone();
-        let _ = self.txs_by_signature.insert(tx_key, Arc::new(tx), self.config.tx_ttl);
+    #[allow(dead_code)]
```
Comment (CjS77) [inline base_layer/core/src/mempool/reorg_pool/reorg_pool.rs]: Can this be deleted now?

File: base_layer/core/src/transaction.rs [hunk 2]
```diff
@@ -514,6 +514,32 @@ impl Transaction {
     pub fn calculate_ave_fee_per_gram(&self) -> f64 {
         (self.body.get_total_fee().0 as f64) / self.calculate_weight() as f64
     }
+
+    /// Returns the minimum maturity of the input UTXOs
+    pub fn min_input_maturity(&self) -> u64 {
+        let mut min_maturity = std::u64::MAX;
+        self.body
+            .inputs
+            .iter()
```
Comment (CjS77) [inline base_layer/core/src/transaction.rs]: self.body.inputs.iter().min() ?

File: base_layer/core/src/transaction.rs [hunk 3]
```diff
@@ -514,6 +514,32 @@ impl Transaction {
...
+        min_maturity
+    }
+
+    /// Returns the maximum maturity of the input UTXOs
+    pub fn max_input_maturity(&self) -> u64 {
+        let mut max_maturity = 0;
+        self.body
+            .inputs
+            .iter()
+            .for_each(|input| max_maturity = max(max_maturity, input.features.maturity));
```
Comment (CjS77) [inline base_layer/core/src/transaction.rs]: self.body.inputs.iter().max() ?

File: base_layer/core/src/transaction.rs (hunk 2 above)
Comment (neonknight64) [reply base_layer/core/src/transaction.rs]: I would have to implement `Ord` for `TransactionInput` that orders them according to `features.maturity` to make it work for `iter().min()` or `iter().max()`. Im not sure this is the right way to order the inputs, but I was able to simplify this code by making use of fold: `       self.body.inputs.iter().fold(std::u64::MAX, |min_maturity, input| {
            min(min_maturity, input.features.maturity)
        })`

File: base_layer/core/src/mempool/reorg_pool/reorg_pool.rs (hunk 1 above)
Comment (neonknight64) [reply base_layer/core/src/mempool/reorg_pool/reorg_pool.rs]: Code removed in next PR.

File: base_layer/core/src/transaction.rs (hunk 3 above)
Comment (neonknight64) [reply base_layer/core/src/transaction.rs]: Code improved with fold.

**Pull request merged**